File: vscode/src/cpp/morphology.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace std;

//画像ファイルの表示
void morpho(char* name)
{
	cv::Mat image = cv::imread(name);	//画像の読み込み
	cv::Mat out = image;

	if (image.empty())	//画像が正しく読み込めたのかを確認
	{
		cout << "画像ファイルを読み込めませんでした．" << endl;
		return;
	}

	cv::imshow("元画像", image);	//画像の表示
	cv::erode(out,out,cv::Mat(),cv::Point(-1,-1),15);
	cv::dilate(out,out,cv::Mat(),cv::Point(-1,-1),120);
	cv::erode(out,out,cv::Mat(),cv::Point(-1,-1),105);
	cv::imshow("結果画像",out);

	while (cv::waitKey() != 'q')
	{
		switch(cv::waitKey()){
			// case 'e':
			// cv::erode(out,out,cv::Mat(),cv::Point(-1,-1),15);
			// cout << "erode" << endl;
			// break;

			// case 'd':
			// cv::dilate(out,out,cv::Mat(),cv::Point(-1,-1),15);
			// cout << "dilate" << endl;
			// break;

			case 's':
			cv::imwrite("../image/result/mask-result.png",out);

		}
		// cv::imshow("結果画像",out);
	}
	
}

int main(int argc, char* argv[])
{
    if (argc < 2)
    {
        morpho("../image/angle1.png");
    }
	else{
        morpho(argv[1]);
    }
	
	return 0;
}

File: python/src/opencv/morphology.py
import sys
import cv2
import numpy as np

class morphology:

    def __init__(self,name):
        self.name = name

    def denoise(self):
        image = cv2.imread(self.name)
        dst = image
        kernel = np.ones((3,3),np.uint8)
        cv2.imshow("元画像",image)

        dst = cv2.erode(dst,kernel,iterations = 15)
        dst = cv2.dilate(dst,kernel,iterations = 120)
        dst = cv2.erode(dst,kernel,iterations = 105)
        cv2.imshow("結果画像",dst)

        if cv2.waitKey() == ord("s"):
            num = self.name.rfind("/")
            str = "./result/ml-" + self.name[num+1:]
            cv2.imwrite(str,dst)
        
        # while cv2.waitKey() != ord("q"):
        #     cv2.imshow("結果画像",dst)
            
        #     if cv2.waitKey() == ord("e"):
        #         dst = cv2.erode(dst,kernel,iterations = 1)

        #     if cv2.waitKey() == ord("d"):
        #         dst = cv2.dilate(dst,kernel,iterations = 1)

        #     if cv2.waitKey() == ord("s"):
        #         num = self.name.rfind("/")
        #         str = "./result/ml-" + self.name[num+1:]
        #         cv2.imwrite(str,dst)
        
        cv2.waitkey(0)
        cv2.destroyAllWindows()

args = sys.argv
if len(args) > 1:
    mp = morphology(args[1])
else:
    mp = morphology("./result/color-angle1.png")

mp.denoise()